pandas/core/style.py

```diff
@@ -0,0 +1,690 @@
+        """
+        Render the built up styles to HTML
+
+        Parameters
+        ----------
+        table_styles: list or None
+            defaults to ``self.table_styles``
+        caption: str or None
+            defaults to ``self.caption``
+        uuid: str or None
+            defaults to random uuid
+        precision: int or None
+            defaults to self.precision
+        trim: bool
+            Small optimization to not create classes for cells that
+            have no styles applied to them.
+
+        Returns
+        -------
+        rendered: str
+            the rendered HTML
+
+        Notes
+        -----
+        ``Styler`` objects have defined the ``_repr_html_`` method
+        which automatically calls ``self.render()`` when it's the
+        last item in a Notebook cell. When calling ``Styler.render()``
+        directly, wrap the resul in ``IPython.display.HTML`` to view
+        the rendered HTML in the notebook.
+        """
+        self._thunk()
+        table_styles = table_styles or self.table_styles
+        caption = caption or self.caption
+        uuid = uuid or self.uuid
+        d = self._translate(table_styles=table_styles, caption=caption,
+                            uuid=uuid)
+        if trim:
+            # filter out empty styles, every cell will have a class
+            # but the list of props may just be [['', '']].
+            # so we have the neested anys below
+            trimmed = [x for x in d['cellstyle'] if
+                       any(any(y) for y in x['props'])]
+            d['cellstyle'] = trimmed
+        return self.template.render(**d)
+
+    def _update_ctx(self, attrs):
+        """
+        update the state of the Styler. Collects a mapping
+        of {index_label: ['<property>: <value>']}
+
+        attrs: Series or DataFrame
+        should contain strings of '<property>: <value>;<prop2>: <val2>'
+        Whitespace shouldn't matter and the final trailing ';' shouldn't
+        matter.
+        """
+        for row_label, v in attrs.iterrows():
+            for col_label, col in v.iteritems():
+                i = self.index.get_indexer([row_label])[0]
+                j = self.columns.get_indexer([col_label])[0]
+                for pair in col.rstrip(";").split(";"):
+                    self.ctx[(i, j)].append(pair)
+
+    def _copy(self, deepcopy=False):
+        styler = Styler(self.data, precision=self.precision,
+                        caption=self.caption, uuid=self.uuid,
+                        table_styles=self.table_styles)
+        if deepcopy:
+            styler.ctx = copy.deepcopy(self.ctx)
+        else:
+            styler.ctx = self.ctx
+        return styler
+
+    def __copy__(self):
+        """
+        Deep copy by default.
+        """
+        return self._copy(deepcopy=False)
+
+    def __deepcopy__(self, memo):
+        return self._copy(deepcopy=True)
+
+    def clear(self):
+        self.ctx.clear()
+
+    def _thunk(self):
```

Comment:
@jreback @jorisvandenbossche here's where I'm doing the lazy evaluation of styles, which lets us share them. As you `.apply` or `.applymap` style functions, they get added to a simple `style._todo` list. The tricky bit not including any reference to `self`, since that will change when you `export` the style (or you don't even have a `self` if we're bundling together some styles from `Styler.higlight_null`. Its understood, *by convention* that the first item in each tuple on `._todo` a `lambda` saying `getattr(instance, 'apply')` or `getattr(instance, 'applymap')` where `instance` is a `style`. Is this OK? Feels a bit hacky, but it works.